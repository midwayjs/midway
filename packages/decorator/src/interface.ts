import { ScopeEnum } from './common/scopeEnum';
import { Middleware } from 'koa';
import { RequestHandler } from 'express';

export type MiddlewareParamArray = Array<Middleware | RequestHandler | string>;
export type ObjectIdentifier = string;

/**
 * 内部管理的属性、json、ref等解析实例存储
 */
export interface IManagedInstance {
  type: string;
  value?: any;
  args?: any;
}

export interface ObjectDefinitionOptions {
  isAsync?: boolean;
  initMethod?: string;
  destroyMethod?: string;
  scope?: ScopeEnum;
  constructorArgs?: IManagedInstance[];
  // 是否自动装配
  isAutowire?: boolean;
  namespace?: string;
  srcPath?: string;
}

export interface TagPropsMetadata {
  key: string | number | symbol;
  value: any;
  args?: any;
}

export interface TagClsMetadata {
  id: string;
  originName: string;
}

export interface ReflectResult {
  [key: string]: TagPropsMetadata[];
}

export enum MSProviderType {
  DUBBO = 'dubbo',
  GRPC = 'gRPC',
  HSF = 'hsf',
}

/**
 * grpc decorator metadata format
 */
export namespace GRPCMetadata {
  export interface ProviderOptions {
    serviceName?: string;
    package?: string;
  }

  export interface ProviderMetadata {
    type: MSProviderType,
    metadata: ProviderOptions
  }
}

export namespace FaaSMetadata {

  interface TriggerCommonOptions {
    /**
     * function name
     */
    functionName?: string;
    /**
     * function description
     */
    description?: string;
    /**
     * function memory size, unit: M
     */
    memorySize?: number;
    /**
     * serverless event name
     */
    name?: string;
    /**
     * function timeout value, unit: seconds
     */
    timeout?: number;
    /**
     * function init timeout, just for aliyun
     */
    initTimeout?: number;
    /**
     * function runtime, nodejs10, nodejs12, nodejs14
     */
    runtime?: string;
    /**
     * invoke concurrency, just for aliyun
     */
    concurrency?: number;
    /**
     * function invoke stage, like env, just for tencent
     */
    stage?: string;
    /**
     * environment variable, key-value
     */
    environment?: any;
    /**
     * function invoke role, just for aliyun
     */
    role?: string;
    /**
     * function publish version, just for aliyun
     */
    version?: string;
  }

  export interface EventTriggerOptions extends TriggerCommonOptions {

  }

  export interface HTTPTriggerOptions extends TriggerCommonOptions  {
    path: string;
    method?: 'get' | 'post' | 'delete' | 'put' | 'head' | 'patch' | 'all';
    middleware?: any[];
  }

  export interface APIGatewayTriggerOptions extends HTTPTriggerOptions  {
  }

  export interface OSTriggerOptions extends TriggerCommonOptions  {
    bucket: string;
    events: string | string[];
    filter?: {
      prefix: string;
      suffix: string;
    };
  }

  export interface LogTriggerOptions extends TriggerCommonOptions  {
    source: string;
    project: string;
    log: string;
    retryTime?: number;
    interval?: number;
  }

  export interface TimerTriggerOptions extends TriggerCommonOptions  {
    type: 'cron' | 'every' | 'interval';
    value: string;
    payload?: string;
  }

  export interface MQTriggerOptions extends TriggerCommonOptions  {
    topic: string;
    tags?: string;
    region?: string;
    strategy?: 'BACKOFF_RETRY' | 'EXPONENTIAL_DECAY_RETRY';
  }

  export interface HSFTriggerOptions extends TriggerCommonOptions  {
  }

  export interface MTopTriggerOptions extends TriggerCommonOptions  {
  }

  export interface CDNTriggerOptions extends TriggerCommonOptions  {
  }

  export type EventTriggerUnionOptions = EventTriggerOptions | HTTPTriggerOptions | APIGatewayTriggerOptions | OSTriggerOptions | CDNTriggerOptions | LogTriggerOptions | TimerTriggerOptions | MQTriggerOptions | HSFTriggerOptions | MTopTriggerOptions;

  export interface TriggerMetadata {
    type: ServerlessTriggerType;
    functionName?: string;
    methodName: string,
    metadata: EventTriggerUnionOptions;
  }

}

export enum MidwayFrameworkType {
  WEB = '@midwayjs/web',
  WEB_KOA = '@midwayjs/koa',
  WEB_EXPRESS = '@midwayjs/express',
  FAAS = '@midwayjs/faas',
  MS_HSF = '',
  MS_GRPC = '@midwayjs/grpc',
  MS_RABBITMQ = '@midwayjs/rabbitmq',
  WS_IO = '@midwayjs/socketio',
  WSS = '',
  SERVERLESS_APP = '@midwayjs/serverless-app',
  CUSTOM = '',
  EMPTY = 'empty',
  LIGHT = 'light',
}

export enum ServerlessTriggerType {
  EVENT = 'event',
  HTTP = 'http',
  API_GATEWAY = 'apigw',
  OS = 'os',
  CDN = 'cdn',
  LOG = 'log',
  TIMER = 'timer',
  MQ = 'mq',
  HSF = 'hsf',
  MTOP = 'mtop',
}
