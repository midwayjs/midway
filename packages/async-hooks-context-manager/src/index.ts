export * from './asyncHooksContextManager';
export * from './asyncLocalStorageContextManager';
export { isSemverGreaterThanOrEqualTo, createContextManager } from './util';
