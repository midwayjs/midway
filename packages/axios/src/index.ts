export { AutoConfiguration as Configuration } from './configuration';
export * from './interface';
export * from './serviceManager';
