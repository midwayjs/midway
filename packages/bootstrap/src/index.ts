export * from './interface';
export { isTypeScriptEnvironment } from './util';
export { Bootstrap, BootstrapStarter } from './bootstrap';
export { ClusterManager } from './manager/cp';
export { AbstractForkManager } from './manager/base';
export { setupStickyMaster } from './sticky';
