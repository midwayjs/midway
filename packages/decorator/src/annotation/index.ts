export * from './configuration';
export * from './inject';
export * from './objectDef';
export * from './priority';
export * from './provide';
export * from './schedule';
export * from './pipeline';
// export * from './validate';
// export * from './rule';
export * from './aspect';
export * from './autoload';
export * from './queue';
export * from './task';
export * from './taskLocal';
