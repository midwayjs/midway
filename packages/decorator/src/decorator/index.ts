// common
export * from './common/index';

// faas
export * from './faas/serverlessTrigger';

// framework
export * from './framework/index';
export * from './framework/schedule';

// microservice
export * from './microservice/consumer';
export * from './microservice/provider';
export * from './microservice/rabbitmqListener';
export * from './microservice/kafkaListener';

// rpc
export * from './rpc/hsf';

// task
export * from './task/queue';
export * from './task/task';
export * from './task/taskLocal';

// web
export * from './web/controller';
export * from './web/paramMapping';
export * from './web/requestMapping';
export * from './web/response';

// ws
export * from './ws/webSocketController';
export * from './ws/webSocketEvent';
