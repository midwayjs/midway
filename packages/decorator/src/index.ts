export * from './interface';
export * from './annotation';
export * from './constant';
export * from './common/decoratorManager';
export * from './common/errMsg';
export * from './common/metadata';
export * from './common/scopeEnum';
export * from './faas/fun';
export * from './faas/handler';
export * from './web/requestMapping';
export * from './web/paramMapping';
export * from './web/controller';
export * from './web/response';
export * from './rpc/hsf';
export * from './framework/config';
export * from './framework/logger';
export * from './framework/plugin';
export * from './framework/app';
export * from './ws/webSocketController';
export * from './ws/webSocketEvent';
export * from './microservice/provider';
export * from './microservice/consumer';
export * from './microservice/rabbitmqListener';
export * from './util/index';
export * from './util/dtoHelper';
