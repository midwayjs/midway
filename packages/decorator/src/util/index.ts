import * as util from 'util';
import * as camelcase from 'camelcase';
import * as crypto from 'crypto';

const ToString = Function.prototype.toString;

function fnBody(fn) {
  return ToString.call(fn)
    .replace(/^[^{]*{\s*/, '')
    .replace(/\s*}[^}]*$/, '');
}

export function isClass(fn) {
  if (typeof fn !== 'function') {
    return false;
  }

  if (/^class[\s{]/.test(ToString.call(fn))) {
    return true;
  }

  // babel.js classCallCheck() & inlined
  const body = fnBody(fn);
  return (
    /classCallCheck\(/.test(body) ||
    /TypeError\("Cannot call a class as a function"\)/.test(body)
  );
}

export function isAsyncFunction(value) {
  return util.types.isAsyncFunction(value);
}

export function isGeneratorFunction(value) {
  return util.types.isGeneratorFunction(value);
}

export function isPromise(value) {
  return util.types.isPromise(value);
}

export function isFunction(value) {
  return typeof value === 'function';
}

export function isObject(value) {
  return value !== null && typeof value === 'object';
}

export function isNumber(value) {
  return typeof value === 'number';
}

export function isProxy(value) {
  return util.types.isProxy(value);
}

export function isMap(value) {
  return util.types.isMap(value);
}

export function isSet(value) {
  return util.types.isSet(value);
}

export function isRegExp(value) {
  return util.types.isRegExp(value);
}

export function isUndefined(value) {
  return value === undefined;
}

export function isNull(value) {
  return value === null;
}

export function isNullOrUndefined(value) {
  return isUndefined(value) || isNull(value);
}

export function sleep(sleepTime = 1000) {
  return new Promise<void>(resolve => {
    setTimeout(() => {
      resolve();
    }, sleepTime);
  });
}

const STRIP_COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/gm;

/**
 * get parameter name from function
 * @param func
 */
export function getParamNames(func): string[] {
  const fnStr = func.toString().replace(STRIP_COMMENTS, '');
  let result = fnStr
    .slice(fnStr.indexOf('(') + 1, fnStr.indexOf(')'))
    .split(',')
    .map(content => {
      return content.trim().replace(/\s?=.*$/, '');
    });

  if (result.length === 1 && result[0] === '') {
    result = [];
  }
  return result;
}

/**
 * 按照框架规则返回类名字
 * @param name 类名称
 */
export function classNamed(name: string) {
  return camelcase(name);
}

/**
 * generate a lightweight random id, enough for ioc container
 */
export function generateRandomId(): string {
  // => f9b327e70bbcf42494ccb28b2d98e00e
  return crypto.randomBytes(16).toString('hex');
}
