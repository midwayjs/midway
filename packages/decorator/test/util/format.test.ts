import { FORMAT } from '../../src'

describe('/test/util/format.test.ts', function () {
  it('should test MS constants', function () {
    expect(FORMAT.MS.ONE_SECOND).toEqual(1000);
    expect(FORMAT.MS.ONE_MINUTE).toEqual(60000);
    expect(FORMAT.MS.ONE_HOUR).toEqual(3600000);
    expect(FORMAT.MS.ONE_DAY).toEqual(86400000);
    expect(FORMAT.MS.ONE_WEEK).toEqual(604800000);
    expect(FORMAT.MS.ONE_YEAR).toEqual(31557600000);
  });

  it('should test crontab', function () {
    expect(FORMAT.CRONTAB).toMatchSnapshot();
  });
});
