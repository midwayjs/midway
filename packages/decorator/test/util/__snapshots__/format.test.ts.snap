// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/util/format.test.ts should test crontab 1`] = `
{
  "EVERY_DAY": "0 0 0 * * *",
  "EVERY_DAY_ONE_FIFTEEN": "0 15 1 * * *",
  "EVERY_DAY_ZERO_FIFTEEN": "0 15 0 * * *",
  "EVERY_HOUR": "0 0 * * * *",
  "EVERY_MINUTE": "0 * * * * *",
  "EVERY_PER_10_MINUTE": "0 */10 * * * *",
  "EVERY_PER_10_SECOND": "*/10 * * * * *",
  "EVERY_PER_30_MINUTE": "0 */30 * * * *",
  "EVERY_PER_30_SECOND": "*/30 * * * * *",
  "EVERY_PER_5_MINUTE": "0 */5 * * * *",
  "EVERY_PER_5_SECOND": "*/5 * * * * *",
  "EVERY_SECOND": "* * * * * *",
}
`;
